packages/world/src/utils/InventoryUtils.sol

```diff
@@ -317,145 +380,188 @@ library InventoryUtils {
     }
   }
 
-  function transferAll(EntityId from, EntityId to) public {
-    require(from != to, "Cannot transfer all to self");
+  function transfer(EntityId from, EntityId to, SlotAmount[] memory slotAmounts)
+    public
+    returns (SlotData[] memory fromSlotData)
+  {
+    fromSlotData = new SlotData[](slotAmounts.length);
 
-    // Occupied slots
-    uint16[] memory slots = Inventory._getOccupiedSlots(from);
+    for (uint256 i = 0; i < slotAmounts.length; i++) {
+      uint16 slotFrom = slotAmounts[i].slot;
+      uint16 amount = slotAmounts[i].amount;
 
-    // Inventory is empty
-    if (slots.length == 0) return;
+      require(amount > 0, "Amount must be greater than 0");
 
-    // Iterate through all from's slots
-    for (uint256 i = 0; i < slots.length; i++) {
-      InventorySlotData memory slotData = InventorySlot._get(from, slots[i]);
+      InventorySlotData memory sourceSlot = InventorySlot._get(from, slotFrom);
+      require(!sourceSlot.objectType.isNull(), "Empty slot");
+      fromSlotData[i] = SlotData(sourceSlot.entityId, sourceSlot.objectType, amount);
 
-      if (slotData.entityId.exists()) {
-        // Handle entities (always amount=1)
-        addEntity(to, slotData.entityId);
+      require(from != to, "Cannot transfer to self");
```

Comment:
we can still move stuff around in your own inventory right? do we have a test for that?